### ISSUE TITLE
Weird message about corrupted double-linked list

### ISSUE THREAD

**Issue by math07** [shinnok/johnny#37]
Something happened in johnny, I don't know under which circumstances, I was doing normal things that I do most of the time. It didn't seem to affect johnny as it seemed to work correctly but this print in the debugging terminal.
EDIT : The message is really long and the formatting was bad when I put it in github, go see http://pastebin.com/w65uFFJ1

**Comment by math07** [comment]
The formatting really sucks http://pastebin.com/w65uFFJ1

**Comment by math07** [comment]
Use this pastebin link, the formatting really sucks in my last message : 
http://pastebin.com/w65uFFJ1

**Comment by math07** [comment]
Okay, here is further information.
Use case : move back and forth between sessions moderatly fast.  load one session, load another one, another one, another one etc... And you'll get an error in HashTypeChecker::start() at the line m_johnOutput.clear() which make the error during the Qt function clear(). Look at those stack calls
![capture d ecran de 2015-06-24 20 40 01](https://cloud.githubusercontent.com/assets/9955284/8344565/48e44d50-1ab1-11e5-994c-da0df0c4d1f0.png)
![capture d ecran de 2015-06-24 20 39 30](https://cloud.githubusercontent.com/assets/9955284/8344568/4ef2b556-1ab1-11e5-8658-0ac6caa30375.png)

**Comment by math07** [comment]
It's kind of weir cause I got it two times in a row, but often I don't get it

**Comment by math07** [comment]
I managed to get "ASSERT failure in QList<T>::operator[]: "index out of range", f" parseJohnAnswer of HashTypeChecker while doing the same use case going back and forth fast between sessions. Maybe it's the same error. I messed up with one of my condition in parsing for robustness, I'll rework it and continue testing Johnny.

**Comment by math07** [comment]
NEW thing which isn't related to other bug :
There is also another(hard to see- I mean you don't see it because the formats are good) bug, the m_johnOutput doesn't always have the same number of characters when going fast. I think it's related to multithreading. At first, I implemented only parseJohnOutput on another thread but when we moved to JohnHandler interface, all methods of HashTypeChecker were on another thread like processOutput(). In some cases, when we go fast, some thread which didn't end yet might be writing to m_johnOutput while another thread is clearing() etc..

**Comment by math07** [comment]
I made a lot of progress regarding a fix for this issue but since it's midnight in Canada I'll go to sleep haha!

**Comment by math07** [comment]
Fixed

**Comment by shinnok** [comment]
So what was the problem here, it's not obvious?

**Comment by math07** [comment]
There were two problems when changing fast between sessions:

1) When we call terminate before show=types has finished to change session file for example,  we can have a weirdly formatted output from john. I messed up my conditions to avoid making my list out of range when output doesn't correspond to the normal format.
So my fix is, correct my conditions which verify that the index of the list is always valid and also don't waist time parsing output when show=types don't return success. Even though it does return success and the output is invalid, it'll no more make index out of bound.

2) There seems to be concurrency with the m_johnOutput variable between threads when doing m_johnOutput.append(XXXX) in the processOutput signal or maybe else where. So I didn't always had the same number of characters in the m_johnOutput for the same files when changing fast between sessions. So I removed this member variable and I use qt method QString johnOutput = readAllStandardOutput() in the parsing method. I guess we should still be careful with this for other future JohnHandler.
